Implemented overviewAxis

rohangoel96 opened the pull request:
# What? Why?

Fix #43  .

Changes proposed in this pull request:
- It is hard to tell where one is on the entire timeline once you zoom in hence an overviewAxis is introduced in this PR.
# Checks
- [x] Single commit and [No merge commits](http://nathanleclaire.com/blog/2014/09/14/dont-be-scared-of-git-rebase/)
- [x] Make sure your commit messages end with a Signed-off-by string (this line
  can be automatically added by git if you run the `git-commit` command with
  the `-s` option).
# Any screenshots or GIFs?

If this is a new visual feature please add a before/after screenshot or gif
here with e.g. [GifGrabber](http://www.gifgrabber.com/).
### Before

![screenshot from 2016-06-01 21-37-56](http://i68.tinypic.com/30rqxwk.gif)
### After

![img](http://i64.tinypic.com/2317a8.gif)
# Notify reviewers

@inodb

rohangoel96 committed:
Implemented overviewAxes

Update:
Use functions instead of global variables
Modularize overviewAxes
More sensible variable names and comments added
Change screenshots as new visual functionality added

Signed-off-by: Rohan Goel <rohangoel0296@gmail.com>

inodb commented on js/lib/d3-timeline.js:
```diff
@@ -1,4 +1,7 @@
 // vim: ts=2 sw=2
+var scrolledX;
+var zoomedWidth;
```
This can be done without a global variable (for reasons why global variables are bad: https://gist.github.com/hallettj/64478). The d3-timeline exposes functions to retrieve local variables. I believe there is already a function called `width()` in d3-timeline that you can call to get the width. You might want to make a function for the `scrolledX` similar to the `width()` function.

inodb commented on js/clinicalTimeline.js:
```diff
@@ -156,6 +158,72 @@ window.clinicalTimeline = (function(){
+      .attr("class", "pos_rectangle")
+      .call(dragRectangle);
+    
+    d3.select(divId+" svg").call(dragChart);
```
As mentioned the `timeline` function is getting really big, maybe we can move some of this overviewAxis to a separate function/file?

rohangoel96 commented:
@inodb, I have updated the PR as per your comments.

inodb commented on js/clinicalTimeline.js:
```diff
@@ -261,6 +283,139 @@ window.clinicalTimeline = (function(){
   /**
+   * Hanles the drawing and panning of the overviewAxis
+   */
+  function handleOverviewAxes() {
```
Name of this function is plural (axes), whereas you prolly want the singular (axis). As far as I understand it draws the overviewAxis, so maybe it makes sense to rename it to drawOverviewAxis or something of that nature.

inodb commented on js/clinicalTimeline.js:
```diff
@@ -261,6 +283,139 @@ window.clinicalTimeline = (function(){
+      .attr("fill", "#ccc");
+
+
+    if(advancedView){
```
`advancedView` is a separate feature, so it's better to separate it from this PR. As a side note I think this part should be outside of `handleOverviewAxis` since you don't want to draw the overviewAxis at all if `advancedView` is false.

inodb commented on js/lib/d3-timeline.js:
```diff
@@ -39,7 +39,8 @@
         showBorderLine = false,
-        showBorderFormat = {marginTop: 25, marginBottom: 0, width: 1, color: colorCycle}
+        showBorderFormat = {marginTop: 25, marginBottom: 0, width: 1, color: colorCycle},
+        scrolledX = null
```
Great! This is looking much cleaner now

inodb commented:
I noticed there's a problem when zooming to a small region:
![small_area_zoom](https://cloud.githubusercontent.com/assets/1334004/15790913/2cb26706-29a2-11e6-9f47-8d0354be45da.gif)

Does seem to be ok in week2 branch:

![small_area_zoom_week2](https://cloud.githubusercontent.com/assets/1334004/15790995/d356a004-29a2-11e6-8a0c-c6d8f0c24358.gif)

rohangoel96 commented:
@inodb , fixed the above mentioned bug. Had to update width variable at in addZoomOptions functions after we changed the width. I will squash the commits once the PR is ready for merge :)

rohangoel96 commented:
Also should I send in the PR for week2 as well? I am working on week3's tasks currently i.e the vertical line we discussed about.

inodb commented on js/clinicalTimeline.js:
```diff
@@ -261,6 +284,121 @@ window.clinicalTimeline = (function(){
+      .tickSize(3)
+      .tickPadding(4);
+
+     var overviewAxisDuplicate = d3.svg.axis().scale(xScaleOverview).orient("top")
```
What is the overviewAxisDuplicate for exactly? Could you add some comments to explain?

inodb commented on js/clinicalTimeline.js:
```diff
@@ -261,6 +284,121 @@ window.clinicalTimeline = (function(){
+   * Hanles the drawing and panning of the overviewAxis
+   */
+  function handleOverviewAxis() {
+    var tickTemp = getTickValues(minDays, maxDays, "months");
```
`tickTemp` what does this variable represent? In general variables named `temp` are frowned upon unless they track temperature. There are a few situations where it can make sense to have it represent "temporary", but it's best to avoid.

inodb commented on js/clinicalTimeline.js:
```diff
@@ -261,6 +284,121 @@ window.clinicalTimeline = (function(){
+      .attr("width", overviewAxisWidth)
+      .attr("class", "overview");
+
+    var xScaleOverview = d3.time.scale()
```
Try to add some comments to explain what `xScaleOverview`, `xScaleOverviewZoomed` and `xScaleRectangle` represent.

inodb commented on index.html:
```diff
@@ -20,10 +20,25 @@
+      color: #ddd;
+    }
+
+    .duplicate .tick text{
```
Could you make this a little more specific? Maybe `.timeline .duplicate .tick`. We should try to prevent accidentally changing other classes. Also not entirely sure what `.duplicate` represents here. Are those the double ticks?

inodb commented on index.html:
```diff
@@ -59,6 +74,9 @@
       fill-opacity: .125;
       shape-rendering: crispEdges;
     }
+    .overview {
```
Here also better to use `.timeline .overview` or something of that nature to prevent clashes with other classes on the cbioportal page.

inodb commented on js/clinicalTimeline.js:
```diff
@@ -261,6 +284,121 @@ window.clinicalTimeline = (function(){
+        .attr("x", overviewX)
+        .attr("y", 27)
+        .attr("fill", "rgba(25,116,255, 0.4)")
+        .attr("class", "pos_rectangle")
```
In the google style guide they recommend to use hyphens instead of underscores: https://google.github.io/styleguide/htmlcssguide.xml#CSS_Style_Rules

rohangoel96 commented:
@inodb , updated and squashed. :)
